src/bt_peer_connection.cpp

```diff
@@ -1230,6 +1139,231 @@ namespace {
+			return;
+		}
+
+		std::shared_ptr<torrent> t = associated_torrent().lock();
+		TORRENT_ASSERT(t);
+
+		auto const& files = t->torrent_file().files();
+
+		span<char const> recv_buffer = m_recv_buffer.get();
+
+		int const header_size = 1 + 32 + 4 + 4 + 4 + 4;
+
+		if (recv_buffer.size() < header_size)
+		{
+			return;
+		}
+
+		const char* ptr = recv_buffer.begin() + 1;
+
+		auto const file_root = sha256_hash(ptr);
+		file_index_t file_index{ -1 };
+		for (file_index_t i : files.file_range())
+		{
+			if (files.root(i) == file_root)
+			{
+				file_index = i;
+				break;
+			}
+		}
+		ptr += sha256_hash::size();
+		int const base = detail::read_int32(ptr);
+		int const index = detail::read_int32(ptr);
+		int const count = detail::read_int32(ptr);
+		int const proof_layers = detail::read_int32(ptr);
+
+		hash_request const hr(file_index, base, index, count, proof_layers);
+
+		if (!validate_hash_request(hr, t->torrent_file().files()))
+		{
+			disconnect(errors::invalid_hashes, operation_t::bittorrent, peer_connection_interface::peer_error);
+			return;
+		}
+
+		// subtract one because the the base layer doesn't count
+		int const proof_hashes = std::max(0
+			, proof_layers - (merkle_num_layers(merkle_num_leafs(count)) - 1));
+
+		if (m_recv_buffer.packet_size() != header_size
+			+ (count + proof_hashes) * int(sha256_hash::size()))
+		{
+			disconnect(errors::invalid_hashes, operation_t::bittorrent, peer_connection_interface::peer_error);
+			return;
+		}
+
+		if (!m_recv_buffer.packet_finished()) return;
+
+		auto new_end = std::remove(m_hash_requests.begin(), m_hash_requests.end(), hr);
+		m_hash_requests.erase(new_end, m_hash_requests.end());
+
+		std::vector<sha256_hash> hashes;
+		while (ptr != recv_buffer.end())
+		{
+			hashes.emplace_back(ptr);
+			ptr += sha256_hash::size();
+		}
+
+#ifndef TORRENT_DISABLE_LOGGING
+		if (should_log(peer_log_alert::incoming_message))
+		{
+			peer_log(peer_log_alert::incoming_message, "HASHES", "%d %d %d %d %d"
+				, static_cast<int>(hr.file), hr.base, hr.index, hr.count, hr.proof_layers);
+		}
+#endif
+
+		if (!t->add_hashes(hr, hashes))
+		{
+			disconnect(errors::invalid_hashes, operation_t::bittorrent, peer_connection_interface::peer_error);
+			return;
+		}
+
+		maybe_send_hash_request();
+	}
+
+	void bt_peer_connection::on_hash_reject(int received)
+	{
+		INVARIANT_CHECK;
+
+		TORRENT_ASSERT(received >= 0);
+		received_bytes(0, received);
+
+		if (!peer_info_struct()->protocol_v2)
+		{
+			disconnect(errors::invalid_message, operation_t::bittorrent);
+			return;
+		}
+
+		if (m_recv_buffer.packet_size() != 1 + 32 + 4 + 4 + 4 + 4)
+		{
+			disconnect(errors::invalid_hash_reject, operation_t::bittorrent, peer_connection_interface::peer_error);
+			return;
+		}
+		if (!m_recv_buffer.packet_finished()) return;
+
+		std::shared_ptr<torrent> t = associated_torrent().lock();
+		TORRENT_ASSERT(t);
+
+		auto files = t->torrent_file().files();
```

Comment:
I think this should be `auto const& files = `, right? otherwise we'll get a copy of `file_storage`.